File: spark-on-angel/core/src/main/scala/com/tencent/angel/spark/client/PSClient.scala
/*
 * Tencent is pleased to support the open source community by making Angel available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */

package com.tencent.angel.spark.client

import com.github.fommil.netlib.F2jBLAS

import com.tencent.angel.spark.context.PSContext
import com.tencent.angel.spark.ops.{Initializer, MatrixOps, SparseRowOps, VectorOps}

/**
 * PSClient is a _instance which contains operations for PSVector &  PSMatrix on the PS nodes.
 * These operations can be called on the Spark driver or executor.
 */
private[spark] class PSClient {
  private[spark] val context = PSContext.instance()
  private[spark] val BLAS = new F2jBLAS

  private[spark] val initOps = new Initializer()
  private[spark] val vectorOps = new VectorOps()
  private[spark] val sparseRowOps = new SparseRowOps()
  private[spark] val matrixOps = new MatrixOps()
}

object PSClient {
  private var _instance: PSClient = _

  private [angel] def instance(): PSClient = {
    if (_instance == null) {
      classOf[PSClient].synchronized {
        if (_instance == null) {
          _instance = new PSClient()
        }
      }
    }
    _instance
  }
}

File: angel-ps/core/src/main/java/com/tencent/angel/master/client/PSClient.java
/*
 * Tencent is pleased to support the open source community by making Angel available.
 * 
 * Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
 * 
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * 
 * https://opensource.org/licenses/BSD-3-Clause
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.tencent.angel.master.client;

import com.google.protobuf.ServiceException;
import com.tencent.angel.common.location.Location;
import com.tencent.angel.ipc.TConnection;
import com.tencent.angel.ipc.TConnectionManager;
import com.tencent.angel.master.app.AMContext;
import com.tencent.angel.protobuf.generated.MasterPSServiceProtos.GetThreadStackRequest;
import com.tencent.angel.ps.PSAttemptId;
import com.tencent.angel.ps.impl.PSProtocol;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import java.io.IOException;

/**
 * Master to ps rpc client, just for get thread stack of ps.
 */
public class PSClient implements PSClientInterface {
  private static final Log LOG = LogFactory.getLog(PSClient.class);
  /**master context*/
  private final AMContext context;

  /**connection from master to the worker*/
  private TConnection connection;

  /**rpc protocol*/
  private PSProtocol ps;

  /**
   * Create a PSClient
   * @param context master context
   * @param attemptId ps attempt id
   * @throws IOException
   */
  public PSClient(AMContext context, PSAttemptId attemptId) throws IOException {
    this.context = context;
    this.connection = TConnectionManager.getConnection(context.getConf());

    Location psLoc =
        context.getParameterServerManager().getParameterServer(attemptId.getPsId())
            .getPSAttempt(attemptId).getLocation();
    LOG.info("psLoc= " + psLoc.toString());
    LOG.info("psLoc.getIp()=    " + psLoc.getIp() + "   " + "          psLoc.getPort()=   "
        + psLoc.getPort());
    this.ps = connection.getPSService(psLoc.getIp(), psLoc.getPort());
  }

  @Override
  public String getThreadStack() throws ServiceException {
    PSProtocol pSProtocol = getPS();
    GetThreadStackRequest request = GetThreadStackRequest.newBuilder().build();
    return pSProtocol.psThreadStack(null, request).getStack();
  }

  private PSProtocol getPS() {
    return ps;
  }

}
